### ISSUE TITLE
Weird thing happening with decimals and abi

### ISSUE THREAD

**Issue by kot-begemot** [ethereum/web3.js#1065]
web3.js version: `1.0.0-beta.21`
nodejs version: `v8.4.0`

**tl;dr**

Looks like order of promises affects their results somehow

**Full story**

So, I am trying to read data from contract using an ABI file

```
var tokenAbi = [
  { "constant": true, "inputs": [], "name": "name", "outputs": [{ "name": "", "type": "string" }], "payable": false, "type": "function" },
  { "constant": true, "inputs": [], "name": "decimals", "outputs": [{ "name": "", "type": "uint8" }], "payable": false, "type": "function" },
  { "constant": true, "inputs": [{ "name": "_owner", "type": "address" }], "name": "balanceOf", "outputs": [{ "name": "balance", "type": "uint256" }], "payable": false, "type": "function" },
  { "constant": true, "inputs": [], "name": "symbol", "outputs": [{ "name": "", "type": "string" }], "payable": false, "type": "function" },
  { "anonymous": false, "inputs": [{ "indexed": true, "name": "_from", "type": "address" }, { "indexed": true, "name": "_to", "type": "address"}, { "indexed": false, "name": "_value", "type": "uint256" }], "name": "Transfer", "type": "event" }
];
```

And here is my code

```
Explorer.prototype.retrieveTokenData = function(tokenAddress) {
  var tokenContract = new web3.eth.Contract(tokenAbi, tokenAddress);
  var newTokenData = { address: tokenAddress };

  // Promises, promises everywhere
  var p1 = tokenContract.methods.name().call();
  var p2 = tokenContract.methods.symbol().call();
  var p3 = tokenContract.methods.decimals().call();

  p1.then(function(result){
    console.log('Got token name ' + result);
  });

  p2.then(function(result){
    console.log('Got token symbol ' + result);
  });

  p3.then(function(result){
    console.log('Got token decimals ' + result);
  });
  // End of promises

  Promise.all([p1, p2, p3]).then(function(arrayOfResults) {
    newTokenData.name = arrayOfResults[0];
    newTokenData.symbol = arrayOfResults[1];
    newTokenData.decimals = arrayOfResults[2];

    // write to database
  });
}
```

Contract address on Ropsten is this `0x95A48dca999c89e4E284930D9b9af973a7481287`

The transaction receipt that I am using is this one

```
 { address: '0x95A48dca999c89e4E284930D9b9af973a7481287',
    topics:
     [ '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
       '0x0000000000000000000000008b0333fa45185a03d4cdc98f6a40eba8a2c393f3',
       '0x00000000000000000000000037e31e7fbd2e3c0b031ccf88c49eb7be54dbb906' ],
    data: '0x0000000000000000000000000000000000000000000000000000000005f5e100',
    blockNumber: 1719097,
    transactionHash: '0xe8dfc366f7dcff72f5110badc5235103c972107fc26d15016fdbb9c71f6b05cb',
    transactionIndex: 0,
    blockHash: '0x18378945072085b3f3b026667b3b75a058a102f17890105d34bdaf3955a8146e',
    logIndex: 0,
    removed: false,
    id: 'log_c3dfa729' }
```

Here is a strange thing, either decimals are returning correct value or name + symbol. It is really weird, i have no idea why this behavior occurs and if that is an issue with my code or with web3.js

It seems to me that the order of promises somehow is important.

**Case 1**
If i declare them in order p1, p2, p3, the output is following

```
New transaction imported 0xe8dfc366f7dcff72f5110badc5235103c972107fc26d15016fdbb9c71f6b05cb
Got token symbol 32
Got token name 32
Got token decimals 8
```

Decimals are returned correctly, however contract name and symbol are returned as numbers '32' 

**Case 2**

Same happens if I declare them in order `p2, p1, p3`  I get the same outcome as in  case 1

**Case 3**

If I declare them in order `p3, p1, p2` I am getting an error for decimals, but name and symbol are correct

```
New transaction imported 0xe8dfc366f7dcff72f5110badc5235103c972107fc26d15016fdbb9c71f6b05cb
Got token name bet
Unhandled rejection Error: The parameter "0x" must be a valid HEX string.
    at Object.hexToUtf8 (/home/ejiao/citowise/token-explorer/node_modules/web3-utils/src/utils.js:186:15)
    at SolidityTypeString.formatOutputString [as _outputFormatter] (/home/ejiao/citowise/token-explorer/node_modules/web3-eth-abi/src/formatters.js:243:31)
    at /home/ejiao/citowise/token-explorer/node_modules/web3-eth-abi/src/type.js:246:25
    at SolidityTypeString.SolidityType.decode (/home/ejiao/citowise/token-explorer/node_modules/web3-eth-abi/src/type.js:247:11)
    at /home/ejiao/citowise/token-explorer/node_modules/web3-eth-abi/src/index.js:327:49
    at Array.forEach (<anonymous>)
    at ABICoder.decodeParameters (/home/ejiao/citowise/token-explorer/node_modules/web3-eth-abi/src/index.js:326:13)
    at Contract._decodeMethodReturn (/home/ejiao/citowise/token-explorer/node_modules/web3-eth-contract/src/index.js:490:22)
    at Method._parent._ethereumCall.call.method.outputFormatter (/home/ejiao/citowise/token-explorer/node_modules/web3-eth-contract/src/index.js:820:42)
    at Method.formatOutput (/home/ejiao/citowise/token-explorer/node_modules/web3-core-method/src/index.js:179:54)
    at sendTxCallback (/home/ejiao/citowise/token-explorer/node_modules/web3-core-method/src/index.js:446:33)
    at Object.<anonymous> (/home/ejiao/citowise/token-explorer/node_modules/web3-core-requestmanager/src/index.js:144:9)
    at /home/ejiao/citowise/token-explorer/node_modules/web3-providers-ws/src/index.js:76:44
    at Array.forEach (<anonymous>)
    at W3CWebSocket.WebsocketProvider.connection.onmessage (/home/ejiao/citowise/token-explorer/node_modules/web3-providers-ws/src/index.js:53:36)
    at W3CWebSocket._dispatchEvent [as dispatchEvent] (/home/ejiao/citowise/token-explorer/node_modules/yaeti/lib/EventTarget.js:107:17)
    at W3CWebSocket.onMessage (/home/ejiao/citowise/token-explorer/node_modules/websocket/lib/W3CWebSocket.js:234:14)
    at WebSocketConnection.<anonymous> (/home/ejiao/citowise/token-explorer/node_modules/websocket/lib/W3CWebSocket.js:205:19)
    at emitOne (events.js:115:13)
    at WebSocketConnection.emit (events.js:210:7)
    at WebSocketConnection.processFrame (/home/ejiao/citowise/token-explorer/node_modules/websocket/lib/WebSocketConnection.js:547:26)
    at /home/ejiao/citowise/token-explorer/node_modules/websocket/lib/WebSocketConnection.js:321:40
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)

(node:30722) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: The parameter "0x" must be a valid HEX string.
(node:30722) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
Got token symbol bet
```

**Case 4**

If I declare them in order `p3, p2, p1` I get the same outcome as in  case 3

**Case 5**

If I declare them in order `p1, p3, p2` I get the same outcome as in  case 3

**Case 6**

If I declare them in order `p2, p3, p1` I get the same outcome as in  case 3

**Comment by kot-begemot** [comment]
I managed to find a workaround: 

After the token data with name + symbol is saved into DB I call a second function that updates decimals. This way I can have all the data, but it comes at the cost of two transactions and unnecessary logic loops. Would be great to see this bug resolved and to remove the unnecessary complexity level.

**Comment by frozeman** [comment]
Thanks this is related to #1069 will be fixed in beta.22.
Please reopen if its still an issue
